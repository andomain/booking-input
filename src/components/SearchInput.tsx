import React, { FC, ChangeEvent, useState, useEffect } from 'react';
import axios, { AxiosResponse } from 'axios';

import { SearchResponse, API_RESPONSE } from '../types';
import SearchResults from './SearchResults';

interface SearchInputProps {
  id: string;
  name: string;
  label: string;
  placeholder: string;
  onSelect: (data: SearchResponse | null) => void;
}

const API_BASE = 'https://www.rentalcars.com';
const MIN_SEARCH_LENGTH = 2;
const API_RESULT_LIMIT = 6;

export const getApi = async (
  searchTerm: string,
  limit: number,
): Promise<AxiosResponse<API_RESPONSE>> =>
  axios.get(
    `${API_BASE}/FTSAutocomplete.do?solrIndex=fts_en&solrRows=${limit}&solrTerm=${searchTerm}`,
  );

const SearchInput: FC<SearchInputProps> = ({
  id,
  name,
  label,
  placeholder,
  onSelect,
}) => {
  const [typedValue, setTypedValue] = useState('');
  const [selectedValue, setSelectedValue] = useState('');
  const [displaySelected, setDisplaySelected] = useState(false);
  const [results, setResults] = useState<null | SearchResponse[]>(null);
  const [error, setError] = useState('');

  useEffect(() => {
    setError('');
    if (typedValue.length >= MIN_SEARCH_LENGTH) {
      getResults(typedValue);
    } else {
      setResults(null);
    }
  }, [typedValue]);

  const getResults = async (
    searchTerm: string,
    limit = API_RESULT_LIMIT,
  ): Promise<void> => {
    const { data, status } = await getApi(searchTerm, limit);

    if (status === 200) {
      if (data.results.numFound > 0) {
        setResults(data.results.docs);
      } else {
        setResults([]);
      }
    } else {
      setError('Something went wrong');
    }
  };

  const typeHandler = (e: ChangeEvent<HTMLInputElement>): void => {
    const {
      target: { value },
    } = e;
    setTypedValue(value);
  };

  const focusInput = () => setDisplaySelected(false);

  const selectHandler = (displayName: string, data: SearchResponse): void => {
    setDisplaySelected(true);
    setSelectedValue(displayName);
    onSelect(data);
  };

  const reset = (): void => {
    setDisplaySelected(false);
    setSelectedValue('');
    onSelect(null);
  };

  return (
    <div className="SearchInput">
      <label className="SearchInput__Label" htmlFor={id}>
        {label}
      </label>
      <input
        className="SearchInput__Input"
        id={id}
        name={name}
        onClick={focusInput}
        onChange={typeHandler}
        type="text"
        placeholder={placeholder}
        value={displaySelected ? selectedValue : typedValue}
      />
      {error && <div className="SearchInput__Error">{error}</div>}
      <SearchResults results={results} onSelect={selectHandler} reset={reset} />
    </div>
  );
};

export default SearchInput;
