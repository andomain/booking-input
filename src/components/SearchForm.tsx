import React, { useState } from 'react'

import SearchInput from './SearchInput';

const SearchForm = () => {
  const [searchTerm, setSearchTerm] = useState('');

  const onChange = () => console.log('Run change handler here')

  const onSubmit = () => console.log('Submit search here');

  return (
    <form className="SearchForm" autoComplete="off" noValidate onSubmit={onSubmit}>
      <SearchInput id="pickup" name="pickup" label="Pick-up location" onChange={onChange} placeholder="city, airport, station, region and district..."/>
      <button type="submit">Search</button>
    </form>
  )
}

export default SearchForm;
