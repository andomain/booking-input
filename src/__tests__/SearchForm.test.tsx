import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import SearchForm from '../components/SearchForm';
import SearchInput from '../components/SearchInput';

describe("<Form />", () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => wrapper = shallow(<SearchForm />));

  it('should render a containing <form />', () => {
    expect(wrapper.find('form').length).toEqual(1);
  });
  
  it('should contain a <SearchInput />', () => {
    expect(wrapper.containsMatchingElement(<SearchInput />)).toEqual(true);
  });
  
  it('should contain a submit button', () => {
    expect(wrapper.find('button').length).toEqual(1);
    expect(wrapper.find('button').type()).toEqual('submit');
  })
})
