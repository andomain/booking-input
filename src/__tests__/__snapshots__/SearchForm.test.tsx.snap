// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchForm /> should render correctly 1`] = `
<SearchForm>
  <form
    autoComplete="off"
    className="SearchForm"
    noValidate={true}
    onSubmit={[Function]}
  >
    <h1
      className="SearchForm__Title"
    >
      Let's find your ideal car
    </h1>
    <SearchInput
      id="pickup"
      label="Pick-up Location"
      name="pickup"
      onSelect={[Function]}
      placeholder="city, airport, station, region and district..."
    >
      <div
        className="SearchInput"
      >
        <label
          className="SearchInput__Label"
          htmlFor="pickup"
        >
          Pick-up Location
        </label>
        <input
          className="SearchInput__Input"
          id="pickup"
          name="pickup"
          onChange={[Function]}
          onClick={[Function]}
          placeholder="city, airport, station, region and district..."
          type="text"
          value=""
        />
        <SearchResults
          onSelect={[Function]}
          reset={[Function]}
          results={null}
        >
          <div
            className="SearchResults"
          >
            <ul />
          </div>
        </SearchResults>
      </div>
    </SearchInput>
    <button
      className="SearchForm__Submit"
      type="submit"
    >
      Search
    </button>
  </form>
</SearchForm>
`;
